File: basic/dfs/parentfinder/ParentFinder.cpp
#include <iostream>
#include <vector>
using namespace std;

int N;
vector<vector<int>> adj;
vector<int> parent;
vector<bool> visited;

void dfs(int node) {
    visited[node] = true;
    for (int next : adj[node]) {
        if (!visited[next]) {
            parent[next] = node;
            dfs(next);
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N;
    adj.resize(N + 1);
    parent.resize(N + 1, 0);
    visited.resize(N + 1, false);

    for (int i = 0; i < N - 1; ++i) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    dfs(1);

    for (int i = 2; i <= N; ++i) {
        cout << parent[i] << '\n';
    }

    return 0;
}

File: basic/dfs/parentfinder/ParentFinder.py
import sys
sys.setrecursionlimit(10 ** 8)
input = sys.stdin.readline

# 전역 변수
N = 0
adj = []
parent = []
visited = []

def dfs(node):
    global visited, parent, adj
    visited[node] = True
    for next_node in adj[node]:
        if not visited[next_node]:
            parent[next_node] = node
            dfs(next_node)

# 입력 처리
N = int(input())
adj = [[] for _ in range(N + 1)]
for _ in range(N - 1):
    a, b = map(int, input().split())
    adj[a].append(b)
    adj[b].append(a)

# 전역 변수 초기화
parent = [0] * (N + 1)
visited = [False] * (N + 1)

# DFS 실행
dfs(1)

# 결과 출력 (최적화된 출력)
output = []
for i in range(2, N + 1):
    output.append(str(parent[i]))
print('\n'.join(output) + '\n')